File: code_blocks/integrations/third-party-integrations/onesignal_1.kt
import android.app.Application
import com.onesignal.OneSignal
import com.onesignal.OSSubscriptionObserver
import com.onesignal.OSSubscriptionStateChanges
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration

class MyApp : Application(), OSSubscriptionObserver {
    override fun onCreate() {
        super.onCreate()
        Purchases.configure(PurchasesConfiguration.Builder(this, "<revenuecat_api_key>").build())

        OneSignal.initWithContext(this)
        OneSignal.setAppId("<onesignal_app_id>")
        OneSignal.addSubscriptionObserver(this)

        OneSignal.getDeviceState()?.userId?.let {
            Purchases.shared.attribution.setOnesignalID(it)
        }
    }

    override fun onOSSubscriptionChanged(state: OSSubscriptionStateChanges) {
        if (!state.from.isSubscribed && state.to.isSubscribed) {
            state.to.userId?.let { Purchases.shared.attribution.setOnesignalID(it) }
        }
    }
}

File: code_blocks/integrations/third-party-integrations/onesignal_1.ts
import OneSignal, { OSSubscriptionChangedEvent } from 'react-native-onesignal';
import Purchases from '@revenuecat/purchases-react-native';

export async function initOneSignalRevenueCat() {
  await Purchases.configure({ apiKey: '<revenuecat_api_key>' });
  OneSignal.setAppId('<onesignal_app_id>');

  OneSignal.addSubscriptionObserver((event: OSSubscriptionChangedEvent) => {
    const userId = event.to.userId;
    if (userId) {
      Purchases.setOnesignalID(userId);
    }
  });

  const state = OneSignal.getDeviceState();
  const userId = state?.userId;
  if (userId) {
    await Purchases.setOnesignalID(userId);
  }
}
